streams/collection.go

```diff
@@ -35,6 +40,25 @@ func NewCollectionFromArray(array interface{}) ICollection {
 	}
 }
 
+// NewCollectionFromMap Creates a new ICollection from the given iterable. Panics if the provided interface is not an iterable or map.
+func NewCollectionFromMap(collection interface{}) ICollection {
```

Comment:
The idea of a `MapEntry` is pretty good, but the way it is currently implemented, you're adding a full iteration over the set in order to map the key and the value, so in cases where only operations of a constant order are schedule would now be of order `N`

I think the best approach would be to create an implementation of `IIterator` which knows how to iterate over a map.